File: Java/037.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Problem037 {

    private static boolean[] isPrimeSieve;
    private static int SIEVE_LIMIT = 1000000; // Sufficient for this problem

    public static void sievePrimes() {
        isPrimeSieve = new boolean[SIEVE_LIMIT + 1];
        Arrays.fill(isPrimeSieve, true);
        isPrimeSieve[0] = false;
        isPrimeSieve[1] = false;
        for (int p = 2; p * p <= SIEVE_LIMIT; p++) {
            if (isPrimeSieve[p]) {
                for (int i = p * p; i <= SIEVE_LIMIT; i += p) {
                    isPrimeSieve[i] = false;
                }
            }
        }
    }

    public static boolean isPrime(int n) {
        if (n < 0 || n > SIEVE_LIMIT) {
            // Fallback for numbers outside sieve range, though not expected for this problem
            if (n <= 1) return false;
            if (n <= 3) return true;
            if (n % 2 == 0 || n % 3 == 0) return false;
            for (int i = 5; i * i <= n; i = i + 6) {
                if (n % i == 0 || n % (i + 2) == 0) {
                    return false;
                }
            }
            return true;
        }
        return isPrimeSieve[n];
    }

    public static boolean isTruncatablePrime(int n) {
        String s = String.valueOf(n);
        int length = s.length();

        // Check left truncations
        for (int i = 1; i < length; i++) {
            if (!isPrime(Integer.parseInt(s.substring(i)))) {
                return false;
            }
        }

        // Check right truncations
        for (int i = 1; i < length; i++) {
            if (!isPrime(Integer.parseInt(s.substring(0, length - i)))) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        sievePrimes();

        long sumOfTruncatablePrimes = 0;
        int count = 0;

        // Start checking from 10, as single-digit primes are not considered truncatable
        for (int i = 10; i < SIEVE_LIMIT; i++) {
            if (isPrime(i) && isTruncatablePrime(i)) {
                sumOfTruncatablePrimes += i;
                count++;
                if (count == 11) {
                    break;
                }
            }
        }
        System.out.println(sumOfTruncatablePrimes);
    }
}

File: Rust/037.rs
use std::collections::HashSet;

fn sieve_of_eratosthenes(limit: usize) -> HashSet<u32> {
    let mut is_prime = vec![true; limit + 1];
    if limit >= 0 { is_prime[0] = false; }
    if limit >= 1 { is_prime[1] = false; }

    for p in 2..=limit {
        if p * p > limit { // Optimization: only iterate up to sqrt(limit)
            break;
        }
        if is_prime[p] {
            for multiple in (p * p..=limit).step_by(p) {
                is_prime[multiple] = false;
            }
        }
    }

    let mut primes = HashSet::new();
    for (i, &prime) in is_prime.iter().enumerate() {
        if prime {
            primes.insert(i as u32);
        }
    }
    primes
}

fn is_truncatable_prime(n: u32, primes_set: &HashSet<u32>) -> bool {
    let s = n.to_string();
    let len = s.len();

    // Check left-to-right truncations
    for i in 1..len {
        let truncated_s = &s[i..];
        if !primes_set.contains(&truncated_s.parse::<u32>().unwrap()) {
            return false;
        }
    }

    // Check right-to-left truncations
    for i in (1..len).rev() {
        let truncated_s = &s[0..i];
        if !primes_set.contains(&truncated_s.parse::<u32>().unwrap()) {
            return false;
        }
    }
    true
}

fn main() {
    let limit = 10_000_000;
    let primes_set = sieve_of_eratosthenes(limit);

    let mut count = 0;
    let mut truncatable_primes_sum: u64 = 0;

    // Start checking from 10, as 2, 3, 5, 7 are explicitly excluded
    for n in 10..limit {
        if primes_set.contains(&n) {
            if is_truncatable_prime(n, &primes_set) {
                count += 1;
                truncatable_primes_sum += n as u64;
                if count == 11 {
                    break;
                }
            }
        }
    }
    println!("{}", truncatable_primes_sum);
}
